instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/Servlet3Advice.java

```diff
@@ -53,39 +53,29 @@ public static void onEnter(
     }
 
     Context attachedContext = tracer().getServerContext(httpServletRequest);
-    if (attachedContext != null) {
+    if (attachedContext != null && tracer().needsRescoping(attachedContext)) {
+      attachedContext =
+          tracer().updateContext(attachedContext, httpServletRequest, mappingResolver, servlet);
+      scope = attachedContext.makeCurrent();
       // We are inside nested servlet/filter/app-server span, don't create new span
-      if (tracer().needsRescoping(attachedContext)) {
-        attachedContext =
-            tracer().updateContext(attachedContext, httpServletRequest, mappingResolver, servlet);
-        scope = attachedContext.makeCurrent();
-        return;
-      }
-
-      // We already have attached context to request but this could have been done by app server
-      // instrumentation, if needed update span with info from current request.
-      Context currentContext = Java8BytecodeBridge.currentContext();
-      Context updatedContext =
-          tracer().updateContext(currentContext, httpServletRequest, mappingResolver, servlet);
-      if (updatedContext != currentContext) {
-        // runOnceUnderAppServer updated context, need to re-scope
-        scope = updatedContext.makeCurrent();
-      }
       return;
     }
 
     Context currentContext = Java8BytecodeBridge.currentContext();
-    if (currentContext != null
-        && Java8BytecodeBridge.spanFromContext(currentContext).isRecording()) {
-      // We already have a span but it was not created by servlet instrumentation.
-      // In case it was created by app server integration we need to update it with info from
-      // current request.
+    if (attachedContext != null
+        || Java8BytecodeBridge.spanFromContext(currentContext).isRecording()) {
```

Comment:
Won't using `isRecording` cause issues when running with sampling?

Reply:
ya, I think `isRecording` should only be used to short-circuit updating a given span's attributes (or name) and probably nothing else, so would support removing this check

Reply:
I'll try to get rid of `isRecording` next week (my initial attempt failed on glassfish and I wasn't able to repro it locally so it requires a bit of debugging). As `isRecording` was already used before this pr I believe we can merge this if there aren't any objections.